### PR TITLE
fix: bound foreground subprocess output

### LINKED ISSUE #9052: Subprocess tools (bash foreground, grep, lake) buffer full stdout/stderr before truncation/pagination
## Problem

Three tool call sites all hit the same root cause: execa's default in-memory buffering of a subprocess's full stdout/stderr, with truncation/pagination only applied *after* the whole thing is already materialized in memory.

- **`src/tools/bash.ts:199`** (`executeForeground()`) — no `buffer: false`/cap at all. `truncate: true` only shortens the *debug log line*, not the actual `result.stdout`/`result.stderr` returned to the caller. The sibling `executeBackground()` in the same file deliberately bounds retained output with `BACKGROUND_OUTPUT_HEAD_CHARS`/`TAIL_CHARS`/`LOG_CAP_CHARS` — the foreground path has no equivalent. **Impact: medium** — the largest per-call exposure of the three (up to execa's ~100MB/stream default, ~500x the background path's caps).
- **`src/tools/grep.ts:139`** (`GrepTool.execute()`) — buffers ripgrep's entire stdout, then materializes every matched line into an array *before* `offset`/`head_limit` pagination is applied. Impact: low (most invocations don't approach the buffer ceiling).
- **`src/tools/lean/direct/lakeCommands.ts:71`** (`executeLake()`) — buffers all of `lake build`/`test`/`exe`'s stdout/stderr; `capOutput()` only trims the already-fully-buffered string to a 4MB cap afterward. Self-documented tradeoff in the code's own comment. Impact: low (gated by a per-workspace mutex, one in-flight buffering invocation at a time).

None are literally unbounded: execa 10.0.0 defaults `maxBuffer` to 100MB per stream and throws past that rather than growing forever — so the real exposure is "up to ~100MB per stream per call," not truly unbounded.

## Why these aren't free fixes

Each would need a real design decision, not a mechanical bound:
- **bash.ts**: bounding foreground output the same way the background path does would change what a large build/test-run command returns to the calling agent.
- **grep.ts**: streaming ripgrep's output and stopping early once `head_limit` is satisfied changes match-counting/offset semantics and error handling for a killed-early process.
- **lakeCommands.ts**: true incremental capping (vs. buffer-then-slice) would need to change how the existing head/tail-preserving `capOutput()` truncation decides what to keep from both ends.

## Ask

Decide per call site whether the execa-default-buffer-then-truncate shape is acceptable as-is (it already has an implicit ~100MB ceiling) or worth a streaming rewrite — starting with `bash.ts` (medium impact, and the one case with *no* app-tuned cap at all today, unlike the other two).

## Triage note

Premature optimization applies here: none of these are confirmed causing real problems today. Revisit opportunistically (e.g. if a session is ever observed OOMing on a large foreground command) rather than proactively.

### PR DESCRIPTION
**Pull request by LionSR**
## Summary

- stream foreground bash stdout and stderr without execa's full in-memory buffers
- retain independent 4,000-character heads and 50,000-character tails with exact omitted-character markers and prior trim semantics
- preserve live output, timeout, cancellation, spawn, and failure formatting, including Unicode split across chunks
- make grep and lake's existing execa 100,000,000-character ceilings explicit while preserving total-count pagination and lake tail diagnostics
- distinguish output-limit failures so grep never paginates partial results or mislabels them as regex errors
- surface maxBuffer failures correctly when execa returns `failed: true` with exit code 0

Closes #9052

## Validation

- focused subprocess suites: 74 passed
- full suite: 7,520 passed, 4 skipped
- full workspace typecheck passed
- full TypeScript lint passed
- `compile:fast` passed
- formatting and patch whitespace checks passed

## Per-call-site policy

- foreground bash: incremental bounded head/tail retention
- grep: drains to completion under an explicit ceiling because totals and pagination require the full result set
- lake: remains serialized buffer-then-cap under an explicit ceiling, preserving its existing 4,194,304-character tail contract

---

> [!NOTE]
> **Medium Risk**
> Changes core subprocess execution and bash tool output shaping used on every shell command; behavior is heavily tested but mistakes could truncate diagnostics or misclassify failures.
> 
> **Overview**
> **Foreground bash** no longer buffers full command output in memory. It runs with `buffer: false`, incrementally keeps a **4k head / 50k tail** per stdout and stderr (aligned with tool-result truncation), inserts exact **characters elided** markers, preserves trim/Unicode chunk semantics, and still forwards chunks to live hooks. Timeout and failure paths use the retained streams; spawn/cancel diagnostics from `executeCommand` still fall back when no stream data arrived.
> 
> **`executeCommand`** gains UTF-8-safe streaming via `StringDecoder`, optional **`maxBuffer`**, and an **`outputLimitExceeded`** flag on `ExecResult` when execa hits max buffer (treated as failure, exit code 2). Buffered mode enforces limits; **`buffer: false`** skips maxBuffer enforcement for streaming consumers.
> 
> **Grep** sets an explicit **100M** `maxBuffer` (pagination still needs the full ripgrep drain) and surfaces a dedicated error when output exceeds that ceiling instead of paginating partial matches or mislabeling regex failures.
> 
> **Lake commands** keep buffer-then-cap behavior but pin the same **100M** process ceiling and tighten failure handling so max-buffer / ambiguous exits are not reported as success with empty stderr.
> 
> Tests and changelog document the bounded-memory behavior and edge cases.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 563cac6f15380312533b3c56a083dad81776ad60. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

## Test organization

 is intentionally separate: it hoists an  mock to exercise synthetic maxBuffer result states (, , exit code 0).  uses real Node subprocesses for mutex, timeout, and retained-tail integration; sharing the hoisted mock would invalidate those tests or require disproportionate module-cache resets.

### CODE REVIEW

File: src/test-kernel/tools/lean/LakeCommandsOutput.vitest.ts [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+      args: ['build'],
+    });
+
+    expect(result).toEqual({
+      exitCode: -1,
+      stdout: 'partial build output',
+      stderr: 'Command failed: stdout maxBuffer exceeded',
+    });
+  });
+});
```
[the remark this answers is not in the document]
Comment (LionSR) [reply src/test-kernel/tools/lean/LakeCommandsOutput.vitest.ts:69]: Keeping this focused file separate because it hoists an `execa` module mock for synthetic maxBuffer result normalization. Merging that mock into `LeanTools.vitest.ts` would contaminate its real Node subprocess integration tests. I documented the rationale in the file header and PR body in 563cac6f15.

File: src/test-kernel/tools/lean/LakeCommandsOutput.vitest.ts [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+import { beforeEach, describe, expect, it, vi } from 'vitest';
+
+const mocks = vi.hoisted(() => ({ execa: vi.fn() }));
+
+vi.mock('execa', () => ({ execa: mocks.execa }));
+
+// Local imports
+import { runLakeCommand } from '@tools/lean/direct/lakeCommands';
+
+describe('runLakeCommand output failures', () => {
```
[the remark this answers is not in the document]
Comment (LionSR) [reply src/test-kernel/tools/lean/LakeCommandsOutput.vitest.ts:18]: The separate suite is the clean isolation boundary for a hoisted `execa` mock; the existing suite deliberately runs real subprocesses. Added an explicit file-header and PR-body rationale in 563cac6f15 rather than introducing module-cache reset machinery.

File: src/tools/bash.ts [hunk 3]
```diff
@@ -203,22 +289,40 @@ export class BashTool extends defineTool({
...
+    const stdout = createBoundedOutputCapture(
+      FOREGROUND_OUTPUT_HEAD_CHARS,
+      FOREGROUND_OUTPUT_TAIL_CHARS,
+    );
+    const stderr = createBoundedOutputCapture(
+      FOREGROUND_OUTPUT_HEAD_CHARS,
+      FOREGROUND_OUTPUT_TAIL_CHARS,
+    );
     const startedAt = Date.now();
     const result = await executeCommand(command, {
```
[the remark this answers is not in the document]
Comment (LionSR) [reply src/tools/bash.ts:301]: Confirmed and pinned in 563cac6f15: with `buffer: false`, execa does not enforce `maxBuffer`. A real regression streams 10,000 characters with `maxBuffer: 64`, verifies success and complete callback delivery, and confirms no output-limit flag. Foreground bash therefore has no hidden execa ceiling.

File: src/test-kernel/tools/BashTool.vitest.ts [hunk 4]
```diff
@@ -282,6 +282,206 @@ describe('BashTool', () => {
...
+        };
+      },
+    );
+
+    const result = await new BashTool().call({ command: 'whitespace-gap' });
+    const output = String(result.output);
+    assert.match(
+      output,
+      /\[\.\.\. 6,002 characters elided from stdout \.\.\.\]/,
+    );
```
[the remark this answers is not in the document]
Comment (LionSR) [reply src/test-kernel/tools/BashTool.vitest.ts:388]: Fixed in 563cac6f15 by deriving the expected count with `(6_002).toLocaleString()`, matching production formatting under the active locale.
